lib/spack/spack/solver/asp.py

```diff
@@ -27,6 +27,12 @@
 except ImportError:
     clingo = None  # type: ignore
     clingo_cffi = False
+except AttributeError:
+    raise RuntimeError(
+        "Clingo installation may be broken or incomplete, "
+        "please verify clingo has been installed correctly"
+        "\n\nClingo does not provide symbol clingo.Symbol"
+    )
```

Comment:
Can you use `raise from` (have to use `except AttributeError as e`) to preserve the stack trace?

Also, is there a way to examine the error more-closely? It seems that it is `import clingo` that is raising the attribute error, so I would have expected it to be caught by `except ImportError`. This message would at least be reasonable as a warning for that catch as well IMO.

It might also help if we ran a specific bootstrapping check with `spack -d`, which should print the full trace.

Reply:
> Can you use raise from (have to use except AttributeError as e) to preserve the stack trace?

Good suggestion, no problem.

>  It seems that it is import clingo that is raising the attribute error

The issue is actually the `clingo.Symbol` attribute reference (`Symbol` is the attribute that cannot be found).

> This message would at least be reasonable as a warning for that catch as well IMO.

I was under the impression we used the import failure as a way to determine whether or not we need to bootstrap clingo and as such we wouldn't consider a lack of importability to be an error and so shouldn't report that to users. I may have over-interpreted here though.

> It might also help if we ran a specific bootstrapping check with spack -d, which should print the full trace.

We would run this when encountering the error? I'm not sure exactly how much utility (if any) that would add to my use case for this change. For context I'm adding this as I went to uninstall clingo to test the bootstrapping phase on Windows, encountered an error where clingo was not fully uninstalled, which effectively bricked my Spack instance, but it was fairly unclear as to why.

Reply:
Also please feel free to suggest a better error/error message to raise, as I feel like the message I wrote is not that much less vague than the error itself.

Reply:
Why don't you avoid `AttributeError` from being raised?

Reply:
Certainly viable, what are you thinking for alternative behavior if we can't load `clingo.Symbol` but can `import clingo`? Seems to me that should still be an error, we can just handle it directly instead of allowing Python to fail for us, but I'm curious if you had a different idea?

Reply:
tbh I don't know what conditions lead to that problem :p

Reply:
The condition is 99% a broken installation of clingo, which leaves an empty directory, which in turn is recognized by Python as a namespace package.

Reply:
@alalazo my use case was for an uninstall, but otherwise exactly right, uninstall did not remove the `clingo` directory and Python loaded it as a namespace module.